File: 0006 Zigzag Conversion/ZigzagConversion.c
char * convert(char * s, int numRows){
    int len = strlen(s);
    if (len <= numRows || numRows < 2) return s;
    
    char *ret = (char *) calloc(len + 1, sizeof(char));
    
    int i = 0;
    for (int r = 0; r < numRows; r++) {
        int index = r;
        int offset0 = 2 * (numRows - r - 1);
        int offset1 = 2 * r;

        ret[i++] = s[index];
        for (; index < len;) {
            if (offset0 > 0) {
                index += offset0;
                if (index < len) {
                    ret[i++] = s[index];
                }
            }
            if (offset1 > 0) {
                index += offset1;
                if (index < len) {
                    ret[i++] = s[index];
                }
            }
        }
    }
    return ret;
}

File: 0006 Zigzag Conversion/ZigzagConversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
 var convert = function(s, numRows) {
    var len = s.length;
    if (len <= numRows || numRows < 2) return s;

    var ret = "";
    for (var r = 0; r < numRows; r++) {
        var index = r;
        var offset0 = 2 * (numRows - r - 1);
        var offset1 = 2 * r;

        ret += s.charAt(index);
        for (; index < len;) {
            if (offset0 > 0) {
                index += offset0;
                if (index < len) {
                    ret += s.charAt(index);
                }
            }
            if (offset1 > 0) {
                index += offset1;
                if (index < len) {
                    ret += s.charAt(index);
                }
            }
        }
    }
    return ret;
};